Adding all Art-Net packets

minorDeveloper opened the pull request:
Adding all packets as defined by the Art-Net 4 specification

Vitineth committed:
Adding first batch of ksy files

These cover with documentation pulled from the spec itself. These are as of yet not tested thoroughly but will gradually be updated with fixes

* artaddress
* arddiagdata
* artipprog
* artipprogreply
* artpoll
* artpollreply

Vitineth committed:
Adding first batch of ksy files

These cover with documentation pulled from the spec itself. These are as of yet not tested thoroughly but will gradually be updated with fixes

* artaddress
* arddiagdata
* artipprog
* artipprogreply
* artpoll
* artpollreply

Vitineth committed:
Merge remote-tracking branch 'origin/vitineth/artnet-ksy' into vitineth/artnet-ksy

Vitineth committed:
Adding verbose output for better logging

Vitineth committed:
Adding four new specification files

Vitineth committed:
Adding titles to all existing files

Vitineth committed:
Two new packet types! 9 more definitions to go

theMinorDeveloper committed:
ArtTodRequest

theMinorDeveloper committed:
Tab fixes

theMinorDeveloper committed:
ArtTodRequest tabs

Vitineth committed:
Rapidly fixing some docs

Vitineth committed:
Adding ArtVlc

Vitineth committed:
Fixing a recommendation from ksc

Vitineth committed:
Adding ArtInput and ArtFirmwareMaster

Vitineth committed:
Removing pre-built html

Vitineth committed:
Adding ArtFirmwareReply

Vitineth committed:
Adding the firmware format even though its not a packet

minorDeveloper committed:
Setting DMX array min and max

Min = 2, max = 512

theMinorDeveloper committed:
Adding ArtTodData

theMinorDeveloper committed:
Adding ArtTodControl

theMinorDeveloper committed:
Adding ArtRdm

theMinorDeveloper committed:
Adding ArtRdmSub

theMinorDeveloper committed:
Migrating RDM UID to custom type

theMinorDeveloper committed:
Merge remote-tracking branch 'origin/vitineth/artnet-ksy' into vitineth/artnet-ksy

Vitineth commented on artnet/ksy/artcommand.ksy:
```diff
@@ -0,0 +1,64 @@
+  
+  These commands shall be transmitted with EstaMan = 0xFFFF
+
+  |Command  |Meaning                                                                                                                      |
+  |---------|-----------------------------------------------------------------------------------------------------------------------------|
+  |SwoutText|This command is used to re-programme the label associated with the ArtPollReply->Swout fields. Syntax: "SwoutText=Playback&" |
+  |SwinText |This command is used to re-programme the label associated with the ArtPollReply->Swin fields. Syntax: "SwinText=Record&"     |
```
This needs to be represented in a different format - tables currently aren't supported with the libraries being used for markdown right now

(I know this was mine, didn't realise it at the time)

Vitineth replied:
![image](https://github.com/About-Time-Technologies/Lighting-Protocols/assets/9435503/24946650-c792-4197-affd-3e37dc92a11d)

Vitineth replied:
@minorDeveloper also need your opinion on this one and the next one (https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1283524136)

Vitineth replied:
Resolving now that tables are supported
![image](https://github.com/About-Time-Technologies/Lighting-Protocols/assets/9435503/4279ca7b-8dd4-4e19-b560-ad4a92907da8)

Vitineth commented on artnet/ksy/artcommand.ksy:
```diff
@@ -0,0 +1,64 @@
+  |SwoutText|This command is used to re-programme the label associated with the ArtPollReply->Swout fields. Syntax: "SwoutText=Playback&" |
+  |SwinText |This command is used to re-programme the label associated with the ArtPollReply->Swin fields. Syntax: "SwinText=Record&"     |
+
+  All:
+    Receive: Application Specific
+    Unicast Transmit: Application Specific
+    Broadcast: Application Specific
```
This applies to all but this renders like trash, i should have picked a better format:

![image](https://github.com/About-Time-Technologies/Lighting-Protocols/assets/9435503/ef19c140-a447-40a2-b850-05c249a0288c)

Vitineth replied:
Could go for

**All**:
* Receive: _Application Specific_
* Unicast Transmit: _Application Specific_
* Broadcast: _Application Specific_

<details>
<summary>Source</summary>

```markdown
**All**:
* Receive: _Application Specific_
* Unicast Transmit: _Application Specific_
* Broadcast: _Application Specific_
```
</details>

minorDeveloper replied:
We don't have markdown support for tables yet do we?

Vitineth replied:
Not yet, I will have a look through the libraries im using and see if its possible to enable it, would definitely simplify things

Vitineth replied:
Tables are now supported (https://github.com/Vitineth/kaitai_struct_compiler/commit/5838d9264bbb06893dfe55a51dffd7b56b2d709c) so will convert all of these to tables

Vitineth commented on artnet/ksy/artcommand.ksy:
```diff
@@ -0,0 +1,64 @@
+# Development nodes
+#   * esta_manufacturer has a misleading description in the spec. It is listed as going Hi Lo, however the description of the fields themselves describe Lo Hi. Based on experimentation, I am pretty sure its Lo Hi however this may need correcting
```
Should probably move this to the notes on the field instead if we are using this as a reference

Vitineth commented on artnet/ksy/artdiagdata.ksy:
```diff
@@ -0,0 +1,39 @@
+    contents: [0x41, 0x72, 0x74, 0x2d, 0x4e, 0x65, 0x74, 0x00]
+    doc: | 
+      Array of 8 characters, the final character is a null termination.
+      Value = ‘A’ ‘r’ ‘t’ ‘-‘ ‘N’ ‘e’ ‘t’ 0x00
```
Should replace these with normal quotes so it doesn't look as weird

Vitineth commented on artnet/ksy/artdiagdata.ksy:
```diff
@@ -0,0 +1,39 @@
+    contents: [0x00, 0x60]
+    doc: OpAddress (transmitted little endian)
+
+  - id: proto_version
+    contents: [0x00, 0x0e]
+    doc: Protocol version (constant 14, transmitted big endian)
```
Turns out this is not done - needs the rest of the packet

Vitineth commented on artnet/ksy/artfirmwaremaster.ksy:
```diff
@@ -0,0 +1,101 @@
+  - size: 20
+    doc: Spare, controller sets to zero, node does not test
+    
+  - id: data
+    size: 512
+    doc: This array contains the firmware or UBEA data block. The order is hi byte first. The interpretation of this data is manufacturer specific. Final packet should be null packed if less than 512 bytes needed.
```
@minorDeveloper can you read this and tell me if you think this field is 512 bytes long or 1,024 bytes long

> `14    Data[512]    Int16    -    This array contains the firmware or UBEA data block. The order is hi byte first. The interpretation of this data is manufacturer specific. Final packet should be null packed if less than 512 bytes needed.`

minorDeveloper replied:
Everything would make me think 1024, except the very last sentence. I feel like that's the most conclusive and we should go with 512 and try contacting AL for clarity.

Vitineth replied:
Cool, happy leaving it with 512 for now and have raised #17 to track it

Vitineth commented on artnet/ksy/artipprog.ksy:
```diff
@@ -0,0 +1,114 @@
+        type: b1
+        doc: |
+          Set to enable DHCP (if set, ignore the lower bits - )
+          Note: the following values are not skipped if set as the values are still present
```
These values are actually ignored now thanks to the conditionals.

@Vitineth are conditionals represented in the patched compiler? if so, how, if not, can they be?

Vitineth replied:
It does not, tracking moved to https://github.com/Vitineth/kaitai_struct_compiler/issues/2 and closing this once the comment is removed

Vitineth commented on artnet/ksy/artpoll.ksy:
```diff
@@ -0,0 +1,115 @@
+    contents: [0x00, 0x0e]
+    doc: Protocol version (constant 14, transmitted big endian)
+
+  - id: flags
+    type: flags
+    doc: Set behaviour of Node
+    if: _io.size >= 13
+
+  - id: diag_priority
+    type: u1
+    enum: priority_codes
+    if: _io.size >= 14
```
`Flags` and `DiagPriority` are not actually optional, should we remove the `if:`?

minorDeveloper replied:
I think we should remove the `if:` if it produces documentation that implies to the reader that this data is optional. If the only impact is a more resilient parser then I'd argue it can remain.

Vitineth replied:
Yeah will definitely remove the if on these, given that any packet less than 14 is a violation of the protocol I don't think there should be any special handling around it unless we see actual cases of packets that do this. Honestly I think I got confused

Vitineth commented on artnet/ksy/artpoll.ksy:
```diff
@@ -0,0 +1,115 @@
+  - id: target_port
+    type: u4be
+    if: _io.size >= 18 and flags.targeted_mode
+
```
Needs `EstaMan` and `Oem`?

Vitineth commented on artnet/ksy/artpoll.ksy:
```diff
@@ -0,0 +1,115 @@
+        type: b1
+        doc: Enable Targeted Mode
+
+      - id: vlc_transmission
+        type: b1
+        doc: Enable VLC transmission
```
Needs clarity - `0=Enable VLC transmission`, `1=Disable VLC transmission`

minorDeveloper replied:
Would using an enum type add clarity here?

Vitineth replied:
Potentially? They are all boolean fields so an enum feels a bit like overkill. I just wish the spec would be consistent on Set means enabled, Clr means disabled but they can't make up their minds. Have updated it to 

```yaml

      - id: targeted_mode
        type: b1
        doc: Enable Targeted Mode

      - id: disable_vlc_transmission
        type: b1
        doc: Disables VLC transmission when set

```

which should match spec but its just annoying its written like this

minorDeveloper replied:
I agree, this is a reasonable solution and makes it clear to the reader

Vitineth commented on artnet/ksy/artpollreply.ksy:
```diff
@@ -0,0 +1,403 @@
+    type: u2be
+    doc: Node's firmware revision number. The controller should only use this field to decide if a firmware udpate should proceed. The convention is that a higher number is a more recent release of firmware
+    
+  - id: net_switch
+    type: u1
+    doc: Bits 14-8 of the 15 bit Port-Address are encoded into the bottom 7 bits of this field. This is used in combination with SubSwitch and SwIn[] or SwOut[] to produce the full universe address
+    
+  - id: sub_switch
+    type: u1
+    doc: Bits 7-4 of the 15 bit Port-Address are encoded into the bottom 4 bits of this field. This is used in combination with NetSwitch and SwIn[] or SwOut[] to produce the full universe address.
```
Should be consistent with the rest of the spec, this should probably be

```yaml

  - id: port_address
    type: u2le
    doc: 15 bit Port-Address to which this packet is destined
```

minorDeveloper replied:
But where are bits 0-3 of the Port-Address stored? In the example used elsewhere the whole Port-Address is in that one variable.

minorDeveloper replied:
Nevermind, that's covered in the next comment. I feel like sticking with the current representation would be most appropriate as in all other instances we have the full Port-Address in one variable. 

We could put 
```
- id: port_address
  type: u2le
  doc: Bits 14-4 of the 15 bit Port-Address. Used in combination with SubSwitch, NetSwitch, SwIn[] or SwOut[] to produce the full universe address.
```

But I think this has the potential to be confused with how this named variable is used elsewhere in the spec.

Vitineth replied:
Good catch, I didn't read that close enough and missed the difference between this and the others. Have reverted my change to keep it how it is and will resolve this thread

Vitineth commented on artnet/ksy/artpollreply.ksy:
```diff
@@ -0,0 +1,403 @@
+    repeat-expr: 4
+    doc: This array defines output status of the node
+    
+  - id: sw_in
+    type: u4be
+    doc: Bits 3-0 of the 15 bit Port-Addressfor each of the 4 possible input ports are encoded into the low nibble.
+    
+  - id: sw_out
+    type: u4be
+    doc: Bits 3-0 of the 15 bit Port-Addressfor each of the 4 possible output ports are encoded into the low nibble.
```
This might be worth replacing with 

```yaml
- id: sw_in
  type: u1
  repeat: expr
  repeat-expr: 4
```

instead of `u4be` to be more representative of the spec

minorDeveloper replied:
I agree

Vitineth commented on artnet/ksy/artpollreply.ksy:
```diff
@@ -0,0 +1,403 @@
+      The Remote inputs are used for remote event triggering or cueing.
+      Bit fields are active high.
+    
+  - size: 3
+    doc: Spare, not used, set to 0
+    if: _io.size >= 200
+    
+  - id: style
+    type: u1
+    enum: style_codes
+    doc: The Style code defines the equipment style of the device.
+    if: _io.size >= 201
+    
+  - id: mac
+    size: 6
+    doc: MAC Address of the node, set to 0 if node cannot supply this information
+    if: _io.size >= 207
```
These are required in the spec, should be have conditionals on these? Can't remember why I did..

Vitineth replied:
As above, removing this as they are required to be spec compliant. Ifs can be added back in if we discover that there are packets in the wild that actually need these

Vitineth commented on artnet/ksy/artpollreply.ksy:
```diff
@@ -0,0 +1,403 @@
+    size: 6
+    doc: RDMnet & LLRP Default Responder UID
+    if: _io.size >= 224
+    
```
Missing `User`, `RefreshRate` and `Filler`

Vitineth commented on artnet/ksy/artpollreply.ksy:
```diff
@@ -0,0 +1,403 @@
+      - id: format
+        type: u1
+        enum: port_type_format
+        doc: The format of data this port interacts with
```
Shouldn't be `u1` should be `b6`

Vitineth commented on artnet/ksy/artrdm.ksy:
```diff
@@ -0,0 +1,86 @@
+    doc: Version of RDM specification supported by the device
+    enum: rdm_versions
+    
+  - size: 1
+    doc: Pad length to match ArtPoll
+    
+  - size: 7
+    doc: Spare transmitted as zero, receivers don't test
```
Im angry these have to be separate but glad they are documented this way

minorDeveloper replied:
It's horrific

Vitineth commented on artnet/ksy/artrdm.ksy:
```diff
@@ -0,0 +1,86 @@
+    type: u1
+    doc: The low 8 bits of the Port-Address that should action this packet
+    
+  - id: rdm_packet
+    size: 256
+    doc: The RDM packet data excluding the DMX StartCode. The maximum packet length is 255 + 2-byte checksum - 1-byte start code = 256 bytes
```
Is this guaranteed to be 256 or can it ever be less? If it can be less, we can use `size-eos: true` instead of `size` which will just read to the end of the stream instead of requiring a size

Vitineth replied:
@minorDeveloper

minorDeveloper replied:
The specification is wonderfully unhelpful. I don't think we'll be certain about the ranges of this packet until I've written the RDM spec itself. For now I'd suggest using `size-eos: true` and `max: 256` to give the following

```
  - id: rdm_packet
    size-eos: true
    doc: The RDM packet data excluding the DMX StartCode...
    valid:
      max: 256
```

@Vitineth if you're happy with this then feel free to resolve

Vitineth replied:
I will need to have a look at this to check that the `valid` block can actually work that way and constrain length. I want to also check a bit more on how `error-eos` works because there's a chance that can suppress errors if it hits the end of the stream unexpectedly but that might only be for delimited structures.

Vitineth replied:
This follows the same issue with ArtTrigger, so will resolve this in the same way as that

Vitineth commented on artnet/ksy/artrdmsub.ksy:
```diff
@@ -0,0 +1,106 @@
+  - size: 4
+    doc: Spare transmit as zero, receviers don't test
+    
+  - id: data
+    size: sub_count * 2
+    doc: The size of this 16-bit big endian data array is sub_count for Set and GetResponse, and 0 for Get and SetResponse commands
```
I don't think size here works out quite right. It does for `Set` and `GetResponse`, however `Get` and `SetResponse` are both 0 which can't be achieved with `sub_count * 2` constrained to not be 0 however I may not be understanding right

Vitineth replied:
@minorDeveloper

minorDeveloper replied:
Hmm, I'm trying to think of a nice way to do this that doesn't involve conditionals. If conditionals are the only option then I would propose the following:
```
- id: data
  size: "command_class == command_classes::set or command_class == command_classes::get_response ? sub_count * 2 : 0"
  doc: ...
```

Enums have been created using values from ANSI_E1-20 Table A-1

@Vitineth what's your opinion on this solution?

Vitineth replied:
If this builds correctly and has the expected behaviour then I'm happy with it - its not the prettiest but it seems to have the right functionality
(sorry it took me so long to get to this, looked at it and then it slipped my mind)

Vitineth commented on artnet/ksy/artsync.ksy:
```diff
@@ -0,0 +1,58 @@
+    type: u1
+    repeat: expr
+    repeat-expr: 2
+    doc: Transmit as zero
```
Should probably just replace this with `size: 2` given that it doesn't matter

Vitineth commented on artnet/ksy/arttoddata.ksy:
```diff
@@ -0,0 +1,122 @@
+    
+  - id: port
+    type: u1
+    doc: Physical port index in a range of 1-4. This number is used with bind_index to identify the physical port that generated the packets. This is done by referencing data in the ArtPollReply with a matching bind_index. Physical Port = (BindIndex - 1)  * ArtPollReply->NumPortsLow + ArtTodDataPort. As most modern gateways will implmenet one universe per ArtPollReply, ArtTodData->Port will usual ...
```
If it is more readable you can split this across multiple lines and use markdown for formatting ie

```yaml
doc: |
    Physical port [....]

    `ArtPollReplyData -> [...]`
```

however this is just a style choice so feel free to close this without change

Vitineth replied:
@minorDeveloper

Vitineth commented on artnet/ksy/arttodrequest.ksy:
```diff
@@ -0,0 +1,66 @@
+    contents: [0x00]
+    doc: 0x00 is TodFull, send the entire TOD
+
+  - id: len_address
+    type: u1
+    doc: The number of entries in Address that are used. Max value is 32
```
Should we enforce this with

```yaml
valid:
    max: 32
```

Vitineth replied:
@minorDeveloper

Vitineth commented on artnet/ksy/arttodrequest.ksy:
```diff
@@ -0,0 +1,66 @@
+    type: u1
+    doc: The number of entries in Address that are used. Max value is 32
+
+  - id: address
+    size: len_address
+    doc: Array of low bytes for the Port-Address of Nodes that must respond. Combined with Net to give full Port-Address
```
Size is technically always 32 here. We can either enforce it or not - I like this because it will make sure the `address` array is always the right size, however not because its slightly against spec. Could potentially add an unnamed field at the end with `size: 32 - len_address` to force a read to the expected end of stream - however this also depends if devices are compliant with always sending the full 32 - thoughts?

Vitineth replied:
@minorDeveloper

minorDeveloper replied:
I lean towards the opinion that our description of the spec should be as accurate as possible, but the codebase should be capable of handling bad implementations correctly.

Vitineth replied:
I think this is a good shout, will leave it with the current size entry but going to add a comment to explain a little bit. If you're happy with the change, resolve this thread

Vitineth commented on artnet/ksy/arttrigger.ksy:
```diff
@@ -0,0 +1,79 @@
+
+  - id: sub_key
+    type: u1
+    doc: The trigger subkey
+
+  - id: data
+    size-eos: true
+    doc: The interpretation of the payload is defined by the Key
```
Should we enforce to 512 as per spec, or leave as `eos` on the chance of bad devices / bad impls?

Vitineth replied:
@minorDeveloper because it links to https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285049110

minorDeveloper replied:
I would suggest the solution presented for artrdm. Enforce 512 as a max, but using `size-eos: true` for flexibility.

@Vitineth resolve if you're happy with this

Vitineth replied:
It turns out this is not supported, when you try and compile to javascript, for example you get

```
validateAttr(data)
arttrigger: /:
	error: can't compare BytesEosType(None,false,None,None) and CalcIntType
```

We could potentially do

```yaml
size: "_io.size - 18 < 512 ? _io.size - 18 : 512"
```

Which will read up to the remaining length of the stream (if less than 512 remaining), or just 512 if its over or equal. In JS this generates the right expression

```js
this.data = this._io.readBytes(((this._io.size - 18) < 512 ? (this._io.size - 18) : 512));
```

Would also need you to double check that the `- 18` is right here.

Vitineth replied:
Note: this also means we need to redo the pipeline to compile to all languages to be safe rather than just HTML as that seems _a lot_ more permissive than the rest

minorDeveloper replied:
Yeah, I had only been testing against HTML before pushing. 

I would agree with the `- 18`

Vitineth replied:
Sweet, updated it to the expression above so this can be resolved, will open an new PR for the workflow update to keep this clean

Vitineth commented on artnet/ksy/artvlc.ksy:
```diff
@@ -0,0 +1,131 @@
+      The sequence number is used to ensure that ArtNzs packets are used in the correct order. 
+      When Art-Net is carried over a medium such as the Internet, it is possible that ArtNzs packets will reach the receiver out of order. 
+
+  - id: start_code
+    type: u1
+    doc: The DMX512 start code of this packet. Must not be Zero or RDM
```
Should enforce as $91_{16}$ as per spec

Vitineth commented on artnet/ksy/artvlc.ksy:
```diff
@@ -0,0 +1,131 @@
+
+  - id: data
+    size: len_data
+    doc: A variable length array of DMX512 lighting data
```
This shouldn't be here?

Vitineth commented on artnet/ksy/artvlc.ksy:
This packet sucks, like what the fuck

Vitineth reviewed (changes requested):
Some minor tweaks, some major fixes required. See comments above, will re-review after changes complete.

Vitineth committed:
Documenting design choices for EstaMan

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285039950

Vitineth committed:
Fixing quotes on Art-Net

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285040010

Vitineth committed:
Finishing artdiagdata

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285040042

Vitineth committed:
Adding EstaMan and Oem

Needs testing due to the added consuming expression. 

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285041924

Vitineth committed:
Adding clarity

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285042019

Vitineth committed:
Fixing port address on ArtPollReply

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285042548

Vitineth committed:
Integer to array for SwIn/Out in ArtPollReply

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285042917

Vitineth committed:
Adding missing fields to ArtPollReply

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285044330

Vitineth committed:
Fixing field length in ArtPollReply

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285045631

Vitineth committed:
Simplifying Aux in ArtSync

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285048063

Vitineth committed:
Mandating VLC magic number in ArtVlc

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285049719

Vitineth committed:
Removing invalid field in ArtVlc

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285049898

Vitineth committed:
Removing note that is now invalid

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285041016

Vitineth committed:
Merge remote-tracking branch 'origin/vitineth/artnet-ksy' into vitineth/artnet-ksy

Vitineth committed:
Reverting change to Port Address on ArtPollReply

Per https://github.com/About-Time-Technologies/Lighting-Protocols/pull/15#discussion_r1285057124

Vitineth committed:
Adding some extra documentation about sizing

minorDeveloper committed:
Setting max value for len_address

minorDeveloper committed:
Port documentation formatting

minorDeveloper committed:
Tabs formatting fix

minorDeveloper committed:
Remove html file

minorDeveloper committed:
Migrating to flexible RDM packet size

minorDeveloper committed:
Enforcing max data size for ArtTrigger

Vitineth committed:
Converting handling rules into tables

Vitineth committed:
Merge remote-tracking branch 'origin/vitineth/artnet-ksy' into vitineth/artnet-ksy

Vitineth committed:
Making flags and diagpriority non-option in ArtPoll

Vitineth committed:
Making spare, style and mac required in ArtPollReply

Vitineth committed:
Handling variable size results in ArtTrigger

Vitineth committed:
Controlling variable length data for ArtRdm

minorDeveloper committed:
Enums and conditional updates for ArtRdmSub

Vitineth reviewed (approved):
Approving (but skeptically) until we have actual data to test with
